File: built-in plugins dsp/mb-woodwind-recorder.cpp
/**
 * MB Recorder
 * Category : instrument
 * Type     : woodwind
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Gentle recorder with sweet medieval tone
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_WOODWIND_RECORDER_H
#define MB_WOODWIND_RECORDER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbWoodwindRecorder : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-woodwind-recorder";
    static constexpr const char* PLUGIN_NAME    = "MB Recorder";
    static constexpr const char* PLUGIN_TYPE    = "woodwind";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float breath = 0.4f;  // range [0, 1]
    float brightness = 0.5f;  // range [0, 1]
    float vibrato = 0.2f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbWoodwindRecorder() = default;
    ~MbWoodwindRecorder() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.breath = std::clamp(params.breath, 0f, 1f);
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.vibrato = std::clamp(params.vibrato, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Recorder
        return input;
    }
};

#endif // MB_WOODWIND_RECORDER_H

File: built-in plugins dsp/mb-woodwind-recorder.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbWoodwindRecorderPlugin: PluginDefinition = {
  id: "mb-woodwind-recorder",
  slug: "mb-woodwind-recorder",
  name: "MB Recorder",
  category: "instrument",
  type: "woodwind" as any,
  version: "1.0.0",
  description: "Gentle recorder with sweet medieval tone",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 0.55 },
    { type: "triangle", detune: 0, gain: 0.35 },
    { type: "noise", detune: 0, gain: 0.1 },
  ],
  envelope: { attack: 0.03, decay: 0.2, sustain: 0.8, release: 0.1 },
  parameters: [
    {
      id: "breath",
      name: "Breath",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "brightness",
      name: "Brightness",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "vibrato",
      name: "Vibrato",
      type: "float",
      defaultValue: 0.2,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { breath: 0.4, brightness: 0.5, vibrato: 0.2, volume: 0.8 },
};

export default MbWoodwindRecorderPlugin;
